multiqc/utils/software_versions.py

```diff
@@ -0,0 +1,256 @@
+                """\
+                <table class="table mqc_versions_table">
+                    <thead>
+                        <tr>
+                            <th> Group </th>
+                            <th> Software </th>
+                            <th> Version  </th>
+                        </tr>
+                    </thead>
+                """
+            )
+        ]
+        for group, tmp_versions in sorted(versions.items()):
+            html.append("<tbody>")
+            for i, (tool, versions) in enumerate(sorted(tmp_versions.items())):
+                versions = list(map(str, versions))
+                html.append(
+                    dedent(
+                        f"""\
+                        <tr>
+                            <td><samp>{group if (i == 0) else ''}</samp></td>
+                            <td><samp>{tool}</samp></td>
+                            <td><samp>{', '.join(versions)}</samp></td>
+                        </tr>
+                        """
+                    )
+                )
+            html.append("</tbody>")
+        html.append("</table>")
+        return "\n".join(html)
+
+
+def update_versions_from_config(config, report):
+    """Update report with software versions from config if provided"""
+    # Parse software version from config if provided
+    versions_from_config = load_versions_from_config(config)
+    for group, softwares in versions_from_config.items():
+        # Try to find if the software is listed among the executed modules.
+        # Unlisted software are still reported in the `Software Versions` section.
+        module = find_matching_module(group, report.modules_output)
+        for software, versions in softwares.items():
+            # Update versions if the software is listed among the executed modules
+            if module is not None and not config.disable_version_detection:
+                for version in versions:
+                    module.add_software_version(str(version), software_name=software)
+
+                # Get the updated software versions from the module
+                versions = module.versions[software]
+
+            # Add updated software versions to the report
+            report.software_versions[group][software] = versions
+
+
+def load_versions_from_config(config):
+    """Try to load software versions from config"""
+    log.debug("Reading software versions from config.software_versions")
+    software_versions_config = getattr(config, "software_versions", defaultdict(lambda: defaultdict(list)))
+    software_versions_config, is_valid = validate_software_versions(software_versions_config)
+
+    if not is_valid:
+        log.error("Software versions loaded config.software_versions is not in a valid format")
+
+    software_versions_file = defaultdict(lambda: defaultdict(list))
+    file_name = config.version_fn_name
+    if not os.path.isfile(file_name):
+        file_name = file_name.replace(".yaml", ".yml")
```

Comment:
We cannot limit this to a specific file / filename, it must be a glob pattern. The idea of this is to allow many files to sit side by side - this means that we will very quickly have filename conflicts. This could be resolved by using subdirectories, but there's no need.

Better to use `*_mqc_version.yml` instead and allow people to have many of these files in one place.

I had envisaged using the standard MultiQC search mechanism to find these files, and then iterating over the resulting list in the same way as a module.

See original comment: https://github.com/ewels/MultiQC/pull/1927#discussion_r1309388032 (easy to get lost in the long review though 👍🏻 )